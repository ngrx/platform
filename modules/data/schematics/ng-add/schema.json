{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxData",
  "title": "NgRx Data Schema",
  "type": "object",
  "properties": {
    "effect": {
      "type": "boolean",
      "default": true,
      "description":
        "When false use the EntityDataModuleWithoutEffects module instead of EntityDataModule."
    },
    "skipPackageJson": {
      "type": "boolean",
      "default": false,
      "description":
        "Do not add @ngrx/data as dependency to package.json (e.g., --skipPackageJson)."
    },
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to the module.",
      "visible": false
    },
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "aliases": ["p"]
    },
    "module": {
      "type": "string",
      "default": "app",
      "description": "Allows specification of the declaring module.",
      "alias": "m",
      "subtype": "filepath"
    },
    "migrateNgrxData": {
      "type": "boolean",
      "default": false,
      "description": "Migrate from ngrx-data, will rename modules.",
      "alias": "migrate"
    },
    "entityConfig": {
      "type": "boolean",
      "default": true,
      "description": "Create the Entity config file"
    }
  },
  "required": []
}
