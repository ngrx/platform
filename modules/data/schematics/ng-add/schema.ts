export interface Schema {
  path?: string;
  effects?: boolean;
  skipPackageJson?: boolean;
  project?: string;
  module?: string;
  migrateNgrxData?: boolean;
  entityConfig?: boolean;
}
