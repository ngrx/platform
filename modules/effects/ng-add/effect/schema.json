{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxEffect",
  "title": "NgRx Effect Options Schema",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the effect.",
      "type": "string"
    },
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the component.",
      "visible": false
    },
    "flat": {
      "type": "boolean",
      "default": true,
      "description": "Flag to indicate if a dir is created."
    },
    "spec": {
      "type": "boolean",
      "default": true,
      "description": "Specifies if a spec file is generated."
    },
    "module": {
      "type": "string",
      "default": "",
      "description": "Allows specification of the declaring module.",
      "alias": "m",
      "subtype": "filepath"
    },
    "root": {
      "type": "boolean",
      "default": false,
      "description":
        "Flag to indicate whether the effects are registered as root."
    },
    "feature": {
      "type": "boolean",
      "default": false,
      "description": "Flag to indicate if part of a feature schematic."
    },
    "group": {
      "type": "boolean",
      "default": false,
      "description": "Group effects file within 'effects' folder",
      "aliases": ["g"]
    }
  },
  "required": ["name"]
}
