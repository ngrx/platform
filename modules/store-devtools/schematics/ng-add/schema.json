{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxRootState",
  "title": "NgRx Root State Management Options Schema",
  "type": "object",
  "properties": {
    "skipPackageJson": {
      "type": "boolean",
      "default": false,
      "description":
        "Do not add @ngrx/store as dependency to package.json (e.g., --skipPackageJson)."
    },
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the state.",
      "visible": false
    },
    "module": {
      "type": "string",
      "default": "app",
      "description": "Allows specification of the declaring module.",
      "alias": "m",
      "subtype": "filepath"
    },
    "maxAge": {
      "type": "number",
      "default": 25,
      "description":
        "number (>1) | 0 - maximum allowed actions to be stored in the history tree. The oldest actions are removed once maxAge is reached. It's critical for performance. 0 is infinite. Default is 25 for performance reasons."
    }
  },
  "required": []
}
