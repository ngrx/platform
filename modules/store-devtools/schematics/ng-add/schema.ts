export interface Schema {
  skipPackageJson?: boolean;
  path?: string;
  project?: string;
  module?: string;
  maxAge?: number;
  autoPause?: boolean;
}
