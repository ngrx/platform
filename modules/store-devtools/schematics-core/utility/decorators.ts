import * as ts from 'typescript';

/**
 * In TS 4.8 the `decorators` are combined with the `modifiers` array.
 * Once we drop support for older versions, we can remove this function
 * and use `ts.getDecorators` directly.
 */
export function getNodeDecorators(node: ts.Node): ts.Decorator[] | undefined {
  return (ts as any).getDecorators?.(node) || node.decorators;
}
