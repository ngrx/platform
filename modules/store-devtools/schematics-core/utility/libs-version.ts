export const platformVersion = '^6.0.0';
