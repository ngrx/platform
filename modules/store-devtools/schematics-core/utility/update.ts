import {
  Rule,
  SchematicContext,
  Tree,
  SchematicsException,
  chain,
} from '@angular-devkit/schematics';
import { NodePackageInstallTask } from '@angular-devkit/schematics/tasks';

export function updatePackage(name: string): Rule {
  return (tree: Tree, context: SchematicContext) => {
    const pkgPath = '/package.json';
    const buffer = tree.read(pkgPath);
    if (buffer == null) {
      throw new SchematicsException('Could not read package.json');
    }
    const content = buffer.toString();
    const pkg = JSON.parse(content);

    if (pkg === null || typeof pkg !== 'object' || Array.isArray(pkg)) {
      throw new SchematicsException('Error reading package.json');
    }

    const dependencyCategories = ['dependencies', 'devDependencies'];

    dependencyCategories.forEach(category => {
      const packageName = `@ngrx/${name}`;

      if (pkg[category] && pkg[category][packageName]) {
        const firstChar = pkg[category][packageName][0];
        const suffix = match(firstChar, '^') || match(firstChar, '~');

        // TODO: remove beta
        pkg[category][packageName] = `${suffix}6.0.0-beta.3`;
      }
    });

    tree.overwrite(pkgPath, JSON.stringify(pkg, null, 2));
    context.addTask(new NodePackageInstallTask());

    return tree;
  };
}

function match(value: string, test: string) {
  return value === test ? test : '';
}
