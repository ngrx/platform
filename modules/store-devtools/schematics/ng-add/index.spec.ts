import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import { getFileContent } from '@schematics/angular/utility/test';
import * as path from 'path';
import { Schema as StoreDevtoolsOptions } from './schema';
import {
  getTestProjectPath,
  createWorkspace,
} from '@ngrx/schematics-core/testing';

describe('Store-Devtools ng-add Schematic', () => {
  const schematicRunner = new SchematicTestRunner(
    '@ngrx/store-devtools',
    path.join(__dirname, '../collection.json')
  );
  const defaultOptions: StoreDevtoolsOptions = {
    skipPackageJson: false,
    project: 'bar',
    module: 'app',
  };

  const projectPath = getTestProjectPath();

  let appTree: UnitTestTree;

  beforeEach(async () => {
    appTree = await createWorkspace(schematicRunner, appTree);
  });

  it('should update package.json', async () => {
    const options = { ...defaultOptions };

    const tree = await schematicRunner
      .runSchematicAsync('ng-add', options, appTree)
      .toPromise();
    const packageJson = JSON.parse(tree.readContent('/package.json'));

    expect(packageJson.dependencies['@ngrx/store-devtools']).toBeDefined();
  });

  it('should skip package.json update', async () => {
    const options = { ...defaultOptions, skipPackageJson: true };

    const tree = await schematicRunner
      .runSchematicAsync('ng-add', options, appTree)
      .toPromise();
    const packageJson = JSON.parse(tree.readContent('/package.json'));

    expect(packageJson.dependencies['@ngrx/store-devtools']).toBeUndefined();
  });

  it('should be provided by default', async () => {
    const options = { ...defaultOptions };

    const tree = await schematicRunner
      .runSchematicAsync('ng-add', options, appTree)
      .toPromise();
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);
    expect(content).toMatch(
      /import { StoreDevtoolsModule } from '@ngrx\/store-devtools';/
    );
    expect(content).toMatch(
      /StoreDevtoolsModule.instrument\({ maxAge: 25, logOnly: environment.production }\)/
    );
  });

  it('should import into a specified module', async () => {
    const options = { ...defaultOptions };

    const tree = await schematicRunner
      .runSchematicAsync('ng-add', options, appTree)
      .toPromise();
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);
    expect(content).toMatch(
      /import { StoreDevtoolsModule } from '@ngrx\/store-devtools';/
    );
  });

  it('should import the environments correctly', async () => {
    const options = { ...defaultOptions, module: 'app.module.ts' };

    const tree = await schematicRunner
      .runSchematicAsync('ng-add', options, appTree)
      .toPromise();
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);
    expect(content).toMatch(
      /import { environment } from '..\/environments\/environment';/
    );
  });

  it('should fail if specified module does not exist', async () => {
    const options = { ...defaultOptions, module: '/src/app/app.moduleXXX.ts' };
    let thrownError: Error | null = null;
    try {
      await schematicRunner.runSchematicAsync('ng-add', options, appTree);
    } catch (err) {
      thrownError = err;
    }
    expect(thrownError).toBeDefined();
  });

  it('should fail if negative maxAges', async () => {
    const options = { ...defaultOptions, maxAge: -4 };

    let thrownError: Error | null = null;
    try {
      await schematicRunner.runSchematicAsync('ng-add', options, appTree);
    } catch (err) {
      thrownError = err;
    }
    expect(thrownError).toBeDefined();
  });

  it('should fail if maxAge of 1', async () => {
    const options = { ...defaultOptions, maxAge: -4 };

    let thrownError: Error | null = null;
    try {
      await schematicRunner.runSchematicAsync('ng-add', options, appTree);
    } catch (err) {
      thrownError = err;
    }
    expect(thrownError).toBeDefined();
  });

  it('should support a custom maxAge', async () => {
    const options = {
      ...defaultOptions,
      name: 'State',
      maxAge: 5,
    };

    const tree = await schematicRunner
      .runSchematicAsync('ng-add', options, appTree)
      .toPromise();
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);
    expect(content).toMatch(/maxAge: 5/);
  });
});
