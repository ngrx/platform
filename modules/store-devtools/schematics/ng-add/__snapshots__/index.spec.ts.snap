// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store-Devtools ng-add Schematic Store Devtools ng-add Schematic for standalone application provides initial setup 1`] = `
"import { ApplicationConfig, provideZoneChangeDetection, isDevMode } from '@angular/core';
import { provideStoreDevtools } from '@ngrx/store-devtools';

export const appConfig: ApplicationConfig = {
  providers: [provideZoneChangeDetection({ eventCoalescing: true }), provideStoreDevtools({ maxAge: 25, logOnly: !isDevMode() })]
};
"
`;
